src/outputFile.ts

```diff
@@ -0,0 +1,69 @@
+import * as fs from 'fs';
+import * as util from 'util';
+import * as path from 'path';
+
+const mkdirAsync = util.promisify(fs.mkdir);
+const openAsync = util.promisify(fs.open);
+const writeAsync = util.promisify(fs.write);
+
+export default class OutputFile {
+    private fileDescriptor: number | undefined;
+    private writePromise: Promise<void> | undefined;
+    private canWrite: boolean = true;
+
+    public constructor(private readonly fileName: string, private readonly directory: string) {}
+
+    /**
+     * Opens the file. Await this method before using.
+     */
+    public async openAsync(): Promise<string | undefined> {
+        try {
+            await mkdirAsync(this.directory);
+        } catch (e) {}
+        return await this.openFileAsync();
+    }
+
+    private openFileAsync = async (index: number = 0) => {
+        if (index === 5) {
+            this.canWrite = false;
+            return;
+        }
+
+        const datePrefix = new Date()
+            .toISOString()
+            .replace(/T/, '_')
+            .replace(/:|-/g, '')
+            .replace(/\..+/, '');
+
+        // Ensure a unique file name
+        const indexFileName = path.join(
+            this.directory,
+            `${datePrefix}_${Date.now()}${index}_${this.fileName}`,
+        );
+
+        try {
+            this.fileDescriptor = await openAsync(indexFileName, 'ax'); // Append, fail if exists
+        } catch (err) {
+            if (err.code === 'EEXIST') {
+                // try the nex index
+                await this.openFileAsync(index + 1);
+            } else {
+                this.canWrite = false;
+            }
+        }
+
+        this.writePromise = Promise.resolve();
+        return indexFileName;
+    };
+
+    public append(content: string) {
+        if (this.canWrite === false) {
+            return;
+        }
+        this.writePromise = this.writePromise!.then(() => {
+            writeAsync(this.fileDescriptor!, content).catch((e: any) => {

```

Comment:
Did you need to return this expression? In order to ensure the logs are written sequentially?

As an alternative, is it worth using a [write stream](https://nodejs.org/api/fs.html#fs_fs_createwritestream_path_options) for this, since that would handle a bunch of the around buffering writes if they're coming in faster then it can write/etc.?

Reply:
Nope, the promise chain will ensure that entries are written to the file in the order that `append` is called.

Write streams might be better for large amounts of data, but I don't think that's a concern for this yet (especially this just being meant as a demo), so don't think its worth the additional complexity around opening/catching errors.